src/libvmmalloc/libvmmalloc.c

```diff
@@ -371,17 +391,199 @@ libvmmalloc_create(const char *dir, size_t size)
+		LOG(3, "clone or remap");
+
+	case 2:
+		LOG(3, "clone the entire pool file");
+
+		if (libvmmalloc_clone() != NULL)
+			break;
+
+		if (Forkopt == 2) {
+			out_log(NULL, 0, NULL, 0, "Error (libvmmalloc): "
+					"pool cloning failed\n");
+			exit(1);
+		}
+		/* cloning failed; fall-thru to remapping */
+
+	case 1:
+		LOG(3, "remap the pool file as private");
+
+		Vmp = mmap(addr, size, PROT_READ|PROT_WRITE,
+				MAP_PRIVATE|MAP_FIXED, Fd, 0);
+
+		if (Vmp == MAP_FAILED) {
+			out_log(NULL, 0, NULL, 0, "Error (libvmmalloc): "
+					"remapping failed\n");
+			exit(1);
+		}
+
+		if (Vmp != addr) {
+			out_log(NULL, 0, NULL, 0, "Error (libvmmalloc): "
+					"wrong address\n");
+			exit(1);
+		}
+
+		Private = 1;
+		break;
+
+	case 0:
+		LOG(3, "do nothing");
+		break;
+
+	default:
+		FATAL("invalid fork action %d", Forkopt);
+	}
+}
+
+/*
+ * libvmmalloc_postfork_parent -- (internal) parent post-fork handler
+ */
+static void
+libvmmalloc_postfork_parent(void)
+{
+	LOG(3, NULL);
+
+	if (Forkopt == 0) {
+		/* do nothing */
+		return;
+	}
+
+	if (Private) {
+		LOG(3, "pool mapped as private - do nothing");
+	} else {
+		LOG(3, "close the cloned pool file");
+		(void) close(Fd_clone);
+	}
+}
+
+/*
+ * libvmmalloc_postfork_child -- (internal) child post-fork handler
+ */
+static void
+libvmmalloc_postfork_child(void)
+{
+	LOG(3, NULL);
+
+	if (Forkopt == 0) {
+		/* do nothing */
+		return;
+	}
+
+	if (Private) {
+		LOG(3, "pool mapped as private - do nothing");
+	} else {
+		LOG(3, "close the original pool file");
+		(void) close(Fd);
+		Fd = Fd_clone;
+
+		void *addr = Vmp->addr;
+		size_t size = Vmp->size;
+
+		LOG(3, "mapping cloned pool file at %p", addr);
+		Vmp = mmap(addr, size, PROT_READ|PROT_WRITE,
+				MAP_SHARED|MAP_FIXED, Fd, 0);
+		if (Vmp == MAP_FAILED) {
+			out_log(NULL, 0, NULL, 0, "Error (libvmmalloc): "
+					"mapping failed\n");
+			exit(1);
+		}
+
+		if (Vmp != addr) {
+			out_log(NULL, 0, NULL, 0, "Error (libvmmalloc): "
+					"wrong address\n");
+			exit(1);
+		}
+	}
+
+	/* XXX - open a new log file, with the new PID in the name */
+}
+
+/*
  * libvmmalloc_init -- load-time initialization for libvmmalloc
  *
  * Called automatically by the run-time loader.
+ * The constructor priority guarantees this is executed before
+ * libjemalloc constructor.
  */
-__attribute__((constructor))
+__attribute__((constructor(101)))
 static void
 libvmmalloc_init(void)
 {
-	char *size_str;
-	char *dir = NULL;
-	size_t size = 0;
+	char *env_str;
+	size_t size;
+
+	/*
+	 * Register fork handlers before jemalloc initialization.
+	 * This provides the correct order of fork handlers execution.
+	 * Note that the first malloc() will trigger jemalloc init, so we
+	 * have to register fork handlers before the call to out_init(),
+	 * as it may indirectly call malloc() when opening the log file.
+	 */
+	if (pthread_atfork(libvmmalloc_prefork,
+			libvmmalloc_postfork_parent,
+			libvmmalloc_postfork_child) != 0) {
+		perror("pthread_atfork");
+		exit(1);
```

Comment:
In this case, if you're using `perror()` because it is too early to use the `out_` routines, you should at least include the library name so there's some indication who caused the program to die.  Something like `perror("libvmmalloc: pthread_atfork");`.  And I think we should uniformly call `abort()` and not `exit()` for these cases since the recent change to `out_fatal()` set us in that direction.